src/common/type/pack.c

```diff
@@ -0,0 +1,1513 @@
+typedef struct PackTypeData
+{
+    PackType type;                                                  // Data type
+    bool valueSingleBit;                                            // Can the value be stored in a single bit (e.g. bool)
+    bool valueMultiBit;                                             // Can the value require multiple bits (e.g. integer)
+    bool size;                                                      // Does the type require a size (e.g. string)
+    const String *const name;                                       // Type name used in error messages
+} PackTypeData;
+
+static const PackTypeData packTypeData[] =
+{
+    {
+        .type = pckTypeUnknown,
+        .name = STRDEF("unknown"),
+    },
+    {
+        .type = pckTypeArray,
+        .name = STRDEF("array"),
+    },
+    {
+        .type = pckTypeBin,
+        .valueSingleBit = true,
+        .size = true,
+        .name = STRDEF("bin"),
+    },
+    {
+        .type = pckTypeBool,
+        .valueSingleBit = true,
+        .name = STRDEF("bool"),
+    },
+    {
+        .type = pckTypeI32,
+        .valueMultiBit = true,
+        .name = STRDEF("i32"),
+    },
+    {
+        .type = pckTypeI64,
+        .valueMultiBit = true,
+        .name = STRDEF("i64"),
+    },
+    {
+        .type = pckTypeObj,
+        .name = STRDEF("obj"),
+    },
+    {
+        .type = pckTypePtr,
+        .valueMultiBit = true,
+        .name = STRDEF("ptr"),
+    },
+    {
+        .type = pckTypeStr,
+        .valueSingleBit = true,
+        .size = true,
+        .name = STRDEF("str"),
+    },
+    {
+        .type = pckTypeTime,
+        .valueMultiBit = true,
+        .name = STRDEF("time"),
+    },
+    {
+        .type = pckTypeU32,
+        .valueMultiBit = true,
+        .name = STRDEF("u32"),
+    },
+    {
+        .type = pckTypeU64,
+        .valueMultiBit = true,
+        .name = STRDEF("u64"),
+    },
+};
+
+/***********************************************************************************************************************************
+Object types
+***********************************************************************************************************************************/
+typedef struct PackTagStack
+{
+    PackType type;
+    unsigned int idLast;
+    unsigned int nullTotal;
+} PackTagStack;
+
+struct PackRead
+{
+    MemContext *memContext;                                         // Mem context
+    IoRead *read;                                                   // Read pack from
+    Buffer *buffer;                                                 // Buffer to contain read data
+    const uint8_t *bufferPtr;                                       // Pointer to buffer
+    size_t bufferPos;                                               // Position in the buffer
+    size_t bufferMax;                                               // Maximum position of buffer
```

Comment:
What do you mean by `Maximum position of buffer`? Is this the position of the last used byte in the buffer (line 262 makes me think that)? Or total size? The way you're using it in line 246 makes me think it is more like bufferLimit? And `of buffer` - do you mean `of the data in the buffer`?

Reply:
Yeah, this is confusing. It is set from bufUsed() so I think it makes sense to just call it bufUsed. Also updated the comment. Done in 84128037a1a.

We can't just used bufUsed because there are two different buffers we might be looking at depending on how the object was constructed, i.e. a buffer passed from the user or an internal buffer used with IoRead.